support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/web/endpoints/OAuth20RevocationEndpointController.java

```diff
@@ -0,0 +1,191 @@
+                }
+                revokeToken(registryToken);
+            }
+        } else {
+            val registryToken = getOAuthConfigurationContext().getTicketRegistry().getTicket(token, OAuth20AccessToken.class);
+
+            if (registryToken == null) {
+                LOGGER.error("Provided token [{}] has not been found in the ticket registry", token);
+            } else {
+                if (!StringUtils.equals(client, registryToken.getClientId())) {
+                    LOGGER.error("Provided token [{}] is not related with the service [{}]", token, client);
+                    return OAuth20Utils.writeError(response, OAuth20Constants.INVALID_REQUEST);
+                }
+                revokeToken(registryToken);
+            }
+        }
+
+        val mv = new ModelAndView(new MappingJackson2JsonView());
+        mv.setStatus(HttpStatus.OK);
+        return mv;
+    }
+
+    /**
+     * Revoke the provided Refresh Token and it's related Access Tokens.
+     *
+     * @param token the token
+     * @return the model and view
+     */
+    private void revokeToken(final OAuth20RefreshToken token) {
+        LOGGER.debug("Revoking token [{}]", token);
+        getOAuthConfigurationContext().getTicketRegistry().deleteTicket(token);
+
+        token.getAccessTokens().forEach(item-> {
+            LOGGER.debug("Revoking Access Token [{}] related to Refresh Token [{}]", item, token);
+            getOAuthConfigurationContext().getTicketRegistry().deleteTicket(item);
+        });
+    }
+
+    /**
+     * Revoke the provided Access Token.
+     *
+     * @param token the token
+     * @return the model and view
+     */
+    private void revokeToken(final OAuth20AccessToken token) {
+        LOGGER.debug("Revoking token [{}]", token);
+        getOAuthConfigurationContext().getTicketRegistry().deleteTicket(token);
+    }
+
+    /**
+     * Verify if the request related token type is supported.
+     *
+     * @param token the token
+     * @return whether the token type is supported
+     */
+    private boolean isAsupportedTokenType(final String token) {
+        return token.startsWith(OAuth20RefreshToken.PREFIX) || token.startsWith(OAuth20AccessToken.PREFIX);
+    }
+
+    /**
+     * Gets registered service by client id.
+     *
+     * @param clientId the client id
+     * @return the registered service by client id
+     */
+    private OAuthRegisteredService getRegisteredServiceByClientId(final String clientId) {
+        return OAuth20Utils.getRegisteredOAuthServiceByClientId(getOAuthConfigurationContext().getServicesManager(), clientId);
+    }
+
+    /**
+     * Verify if the request related service must be authenticated.
+     *
+     * @param registeredService the registered service
+     * @return whether the service need authentication
+     */
+    private boolean isServiceNeedAuthentication(final OAuthRegisteredService registeredService) {
+        return !StringUtils.isBlank(registeredService.getClientSecret());
```

Comment:
Duplicate code? Move to OAuthUtils, etc. And switch to isNotBlank

Reply:
Done.